### PR TITLE
Add createImageFromTextureSource method to ui_web

### PR DESCRIPTION
**Pull request by jezell**
Adds createImageFromTextureSource for flutter web, exposed in dart:web_ui.

```dart
final ui.Image uiImage = renderer.createImageFromTextureSource(bitmap,
            width: 150, height: 150, transferOwnership: false);
```

In canvaskit renderer, this will use MakeLazyImageFromTextureSource. In SkWasmRenderer, it will call imageCreateFromTextureSource, which was already implemented in SkWasm for createImageFromImageBitmap to use, but was not exposed in a way that it could be taken advantage of.

By default, createImageFromTextureSource will create a copy of the object it is passed, but transferOwnership: true may be specified to allow transferable objects to be transferred to the renderer, avoiding the copy.

Fixes: https://github.com/flutter/flutter/issues/150479
Fixes: https://github.com/flutter/flutter/issues/144815

### CODE REVIEW

**Comment by renancaraujo** [comment]
legend

**Comment by kevmoo** [comment]
@jezell – look at a rebase and failing tests?

**Comment by kevmoo** [comment]
@jezell Are the impeller changes related?

File: DEPS [hunk 1]
```diff
@@ -62,7 +62,7 @@ vars = {
   # Dart is: https://github.com/dart-lang/sdk/blob/main/DEPS
   # You can use //tools/dart/create_updated_flutter_deps.py to produce
   # updated revision list of existing dependencies.
-  'dart_revision': 'bb65648e20e29abc47acf3dd984518d29fd625c3',
+  'dart_revision': '1a849f91804e73a830844a5bf546a3f7027fc36d',
```
Comment (kevmoo) [inline DEPS:65]: Revert this?

File: impeller/tessellator/tessellator.cc [hunk 2]
```diff
@@ -187,6 +187,9 @@ std::vector<Point> Tessellator::TessellateConvex(const Path& path,
                           [this](Path::Polyline::PointBufferPtr point_buffer) {
                             point_buffer_ = std::move(point_buffer);
                           });
+  if (polyline.points->size() == 0) {
```
Comment (kevmoo) [inline impeller/tessellator/tessellator.cc:190]: unrelated, right?

File: lib/web_ui/lib/ui_web/src/ui_web/images.dart [hunk 3]
```diff
@@ -44,3 +44,10 @@ FutureOr<ui.Image> createImageFromImageBitmap(JSAny imageSource) {
     imageSource as DomImageBitmap,
   );
 }
+
+/// Creates a [ui.Image] from a valid texture source (for example HTMLImageElement | HTMLVideoElement | HTMLCanvasElement).
```
Comment (eyebrowsoffire) [inline lib/web_ui/lib/ui_web/src/ui_web/images.dart:48]: The main issue I have here is defining clear ownership semantics. `HTMLImageElement`, `HTMLVideoElement` and `HTMLCanvasElement` are not transferable objects, so this sort of loosely implies that the engine isn't taking ownership of the object. However, the skwasm implementation of this *does* try to take ownership and transfer this object to the raster thread (it doesn't actually do anything different from the above function).

If we want this function to work in a non-owning manner, the skwasm renderer will have to create an image bitmap out of the texture source synchronously and then transfer/use that for the rest of its pipeline.

If we want this function to work in a manner in which the engine takes ownership, it isn't really essentially any different from `createImageFromImageBitmap` except that we're loosening the restriction on the type being passed (i.e. the `domInstanceOfString(imageSource, 'ImageBitmap')` check).

I think there is room for both owning and non-owning APIs. We need to make a decision here on what we want this to do and document it clearly (and treat it consistently with both renderers).

**Review by eyebrowsoffire — CHANGES_REQUESTED** [review]
This looks like it has pulled in a bunch of unrelated changes. Can you clean this PR up so that only contains the changes related to the proposed `createImageFromTextureSource` API?

File: impeller/tessellator/tessellator.cc (hunk 2 above)
Comment (jezell) [reply impeller/tessellator/tessellator.cc:190]: @eyebrowsoffire @kevmoo  I didn't make any of these changes. Was a bit confusing to me that Github decided to include them as well. Maybe because I committed the work on top of the 3.22.2 branch instead of main? I can reapply the changes on top of main instead.

File: lib/web_ui/lib/ui_web/src/ui_web/images.dart (hunk 3 above)
Comment (jezell) [reply lib/web_ui/lib/ui_web/src/ui_web/images.dart:48]: I get the ownership transfer problem, but does there need to be two APIs? There aren't two APIs for creating a VideoFrame in the browser for example. If it transfers, once it transfers, it transfers. Since this is a web specific API that is pretty low level, seems like it should be the caller's responsibility to use transferable objects and not touch them again if they are running in SkWasm? In the test for example, there is a different code path for SkWasm that just uses an ImageBitmap instead. Trying to be too prescriptive about the call could limit it's usefulness in practice and take some control out of the caller's hands.

The existing createImageFromBitmapSource has the same problem and doesn't really do anything special about it, why is this API any different? 

Maybe it's purely a documentation problem? It's probably correct to assume that a lot of people won't about the transferability problem if it's not stated clearly in the docs. I don't think there's a good way to check that something is transferable at the JS level that is future proof and throw a better error message in the SKWasm path, but maybe an assert against the known types of transferable objects on the skwasm path plus some info in the docs could help guide people toward using the API properly?

File: impeller/tessellator/tessellator.cc (hunk 2 above)
Comment (eyebrowsoffire) [reply impeller/tessellator/tessellator.cc]: Yes, I think you'll have to rebase on main to clean this up

File: lib/web_ui/lib/ui_web/src/ui_web/images.dart (hunk 3 above)
Comment (eyebrowsoffire) [reply lib/web_ui/lib/ui_web/src/ui_web/images.dart:48]: The `createImageFromImageBitmap` API specifically says (in its doc comment) that it takes ownership of the object. The caller must assume that the object they passed in is invalid after the call. This is not a "problem", it is just a documented behavior of the API.

It seems like the main impetus behind this change is that you can pass DOM elements (`HTMLImageElement`, `HTMLVideoElement`, `HTMLCanvasElement`, etc) into the API without requiring the user to do a copy step to make it an `ImageBitmap` first. In this case, the API does *not* take ownership of the passed object. The contents of the texture source are expected to be synchronously copied into some internal representation in the renderer, and the caller can expect the source to be valid after the call is over. In the CanvasKit case, it is immediately calling `texImage2D` on the texture source to make a WebGL texture out of it that Skia can use and own internally. The Skwasm side is similar, but the `texImage2D` call actually happens asynchronously on the raster thread. This means in order to have the same semantics, Skwasm must make a copy of the texture source synchronously (via `createImageBitmap` or similar) so that it is (1) transferable and (2) guaranteed to still be around and have the expected contents by the time the web worker starts using it.

The API must have consistent semantics between renderers. If the API accepts an `HTMLVideoElement` when the renderer is CanvasKit, but that same code breaks when using Skwasm, we haven't implemented the API correctly. Users should not have to be aware of which renderer is being used and change their behavior based on that.

Comment (jezell) [reply lib/web_ui/lib/ui_web/src/ui_web/images.dart:48]: @eyebrowsoffire so you think it should maybe create an image bitmap in SkWasm if the object it's passed isn't transferable? Then it's just on the user to take a faster path and send a VideoFrame if they don't want the copy to happen. Easy enough to do with some type checking so that bit can be handled transparently.

Comment (eyebrowsoffire) [reply lib/web_ui/lib/ui_web/src/ui_web/images.dart:48]: I think it should *unconditionally* create an image bitmap copy in skwasm, regardless of whether it's transferable. The semantics of this API should be that the caller can expect the texture source they passed in to be valid after the call. So even if it is transferable, we should make our own copy so as not to invalidate the texture source the user passed in.

If we need an API where people can efficiently pass objects they don't want ownership of anymore, we should just loosen the restrictions of the `createImageFromImageBitmap` API and possibly rename it to make it more general, since it already has that contract.

Alternatively, maybe this API could just take a boolean argument that's something like `transferOwnership` or something like that, so that the renderer can know whether it's safe to use the object directly or whether it needs to make a copy. Then this API could essentially replace `createImageFromImageBitmap`. What do you think?

Comment (jezell) [reply lib/web_ui/lib/ui_web/src/ui_web/images.dart:48]: I like the alternate option of putting the user in control of the transfer. I think that addresses both of our concerns.

File: lib/web_ui/lib/ui_web/src/ui_web/images.dart [hunk 4]
```diff
@@ -45,9 +45,16 @@ FutureOr<ui.Image> createImageFromImageBitmap(JSAny imageSource) {
...
 }
 
-/// Creates a [ui.Image] from a valid texture source (for example HTMLImageElement | HTMLVideoElement | HTMLCanvasElement).
-ui.Image createImageFromTextureSource(Object object,  { required int width, required int height }) {
+/// Creates a [ui.Image] from a valid texture source (for example
+/// HTMLImageElement | HTMLVideoElement | HTMLCanvasElement).
+///
+/// By default, the ownership of the texture will be not be transferred to the
+/// renderer, and a copy of the texture source will be made. If this is not
+/// desired, the ownership of the object can be transferred to the renderer and
```
Comment (kevmoo) [inline lib/web_ui/lib/ui_web/src/ui_web/images.dart:53]: Mention the argument `[transferOwnership]` here!

File: lib/web_ui/lib/ui_web/src/ui_web/images.dart (hunk 3 above)
Comment (jezell) [reply lib/web_ui/lib/ui_web/src/ui_web/images.dart]: @eyebrowsoffire added transferOwnership arg with default of false, which will make a copy

File: lib/web_ui/lib/src/engine/skwasm/skwasm_impl/renderer.dart [hunk 5]
```diff
@@ -472,6 +472,19 @@ class SkwasmRenderer implements Renderer {
       surface.handle,
     ));
   }
+
+  @override
+  FutureOr<ui.Image> createImageFromTextureSource(Object textureSource, { required int width, required int height, required bool transferOwnership }) async {
+    if(!transferOwnership) {
```
Comment (eyebrowsoffire) [inline lib/web_ui/lib/src/engine/skwasm/skwasm_impl/renderer.dart:478]: A space between the `if` and the conditional for consistency with our style guide: https://github.com/flutter/flutter/blob/master/docs/contributing/Style-guide-for-Flutter-repo.md

```suggestion
    if (!transferOwnership) {
```

File: lib/web_ui/lib/src/engine/canvaskit/renderer.dart [hunk 6]
```diff
@@ -238,6 +238,29 @@ class CanvasKitRenderer implements Renderer {
     return CkImage(skImage);
   }
 
+  @override
+  FutureOr<ui.Image> createImageFromTextureSource(Object object,
+      {required int width, required int height, required bool transferOwnership}) async {
+        if(!transferOwnership) {
```
Comment (eyebrowsoffire) [inline lib/web_ui/lib/src/engine/canvaskit/renderer.dart:244]:
```suggestion
        if (!transferOwnership) {
```
Here as well

File: lib/web_ui/test/ui/image_golden_test.dart [hunk 7]
```diff
@@ -106,14 +104,19 @@ Future<void> testMain() async {
 
         final ui.PictureRecorder recorder = ui.PictureRecorder();
         final ui.Canvas canvas = ui.Canvas(recorder, drawRegion);
-        canvas.drawImage(image, ui.Offset.zero, ui.Paint()..filterQuality = ui.FilterQuality.none);
-        canvas.drawImage(image, const ui.Offset(150, 0), ui.Paint()..filterQuality = ui.FilterQuality.low);
-        canvas.drawImage(image, const ui.Offset(0, 150), ui.Paint()..filterQuality = ui.FilterQuality.medium);
-        canvas.drawImage(image, const ui.Offset(150, 150), ui.Paint()..filterQuality = ui.FilterQuality.high);
+        canvas.drawImage(image, ui.Offset.zero,
```
Comment (eyebrowsoffire) [inline lib/web_ui/test/ui/image_golden_test.dart:107]: There are a ton of changes in this file that are just random whitespace differences. To avoid churn, can you try to revert the changes that are unrelated to the test you added?

File: lib/web_ui/lib/ui_web/src/ui_web/images.dart [hunk 8]
```diff
@@ -44,3 +44,17 @@ FutureOr<ui.Image> createImageFromImageBitmap(JSAny imageSource) {
...
+
+/// Creates a [ui.Image] from a valid texture source (for example
+/// HTMLImageElement | HTMLVideoElement | HTMLCanvasElement).
+///
+/// By default, [transferOwnership] specifies that the ownership of the texture
+/// will be not be transferred to the renderer, and a copy of the texture source
+/// will be made. If this is not desired, the ownership of the object can be
+/// transferred to the renderer and the engine will take ownership of the
+/// texture source and consume its contents.
+FutureOr<ui.Image> createImageFromTextureSource(Object object,  { required int width, required int height, bool transferOwnership = false }) {
```
Comment (eyebrowsoffire) [inline lib/web_ui/lib/ui_web/src/ui_web/images.dart:56]: I think this should actually take a `JSAny` as a parameter instead of a plain `Object`. There are some subtle differences there on the dart2wasm side and `JSAny` is going to be more correct and less error-prone.

**Review by eyebrowsoffire — APPROVED** [review]
Looks good to me, modulo a few random formatting things, and one requested change to the API surface. Thanks for the fix! I think this will be a good addition.

File: lib/web_ui/lib/ui_web/src/ui_web/images.dart (hunk 8 above)
Comment (kevmoo) [reply lib/web_ui/lib/ui_web/src/ui_web/images.dart:56]: or `JSObject`?

File: lib/web_ui/test/ui/image_golden_test.dart (hunk 7 above)
Comment (jezell) [reply lib/web_ui/test/ui/image_golden_test.dart:107]: Yeah I was kind of fighting the default dart format here. Does the engine have a set of format options or is everyone doing it manual?

Comment (kevmoo) [reply lib/web_ui/test/ui/image_golden_test.dart:107]: We're (thankfully) working on getting formatting rolling into the engine (and hopefully flutter/flutter) after not too long here. Send happy vibes to @matanlurey who is driving!

Comment (kevmoo) [reply lib/web_ui/test/ui/image_golden_test.dart:107]: ...at the moment it's all by hand (in my opinion, silliness)

**Comment by eyebrowsoffire** [comment]
I approved the new goldens.

**Comment by kevmoo** [comment]
Okay to `auto-submit` @eyebrowsoffire ?

**Review by kevmoo — APPROVED** [review]

**Comment by jezell** [comment]
Thanks @kevmoo @eyebrowsoffire
